azurelinuxagent/common/osutil/default.py

```diff
@@ -713,6 +703,34 @@ def get_first_if(self):
 
         return iface.decode('latin-1'), socket.inet_ntoa(sock[i+20:i+24])
 
+    def _get_interface_list_as_string(self, num_devices, struct_size):
```

Comment:
This is an odd method.  I could have made it return a list of interfaces and IP addresses, but that's not really what the original code did.  I ended up keeping it as a "string" and left it to the caller to decode. I am open to suggestions.

Reply:
I'd be happier if you returned a dictionary of interface names mapped the remainder of the string. The Linux struct ifreq is also a weird thing, since aside from the interface name, the rest is a union of possible representations for the interface information, and you have to know what kind of interface it is before you can correctly decode it. It's a sockaddr, and it's probably an IPv4 sockaddr, but that may change in the future, and there's no way to tell.

Reply:
I inherited this code, which makes me a little leery of change.  I was going to go the dictionary route, but I have little confidence in the consequences of those decisions.  If you have confidence, I am happy to take yours.

  1. If I return a dictionary of interface to IP address is there a case where collision will happen?
  1. If I return a dictionary of interface to IP address is there a case where inet_ntoa will throw?

The code is a little odd because the for loop's variable is used outside the scope of the for loop.  Only when the loop is exhausted is the IP address captured.  It wasn't clear if there was intent behind this *oddity* or not.

Reply:
Since you split this out to fix a testing problem, no good deed shall go unpunished, I agree with @jasonzio that it would be nicer if the new method returned a dict. But yes this requires more fiddling in the client code, i.e. where the new method if called.

Reply:
Speaking of no good deed going unpunished... :-) I will prepare a separate PR which does just this one thing. Once we approve/merge, you'll want to unwind this part.

Reply:
Done, and merged to master.